applications/LinearSolversApplication/custom_solvers/amgcl_raw_solver.cpp

```diff
@@ -0,0 +1,748 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+#ifndef AMGCL_PARAM_UNKNOWN
+    #include "input_output/logger.h"
+    #define AMGCL_PARAM_UNKNOWN(NAME)                       \
+        KRATOS_ERROR                                        \
+            << KRATOS_CODE_LOCATION                         \
+            << "Unknown parameter " << NAME << std::endl
+#endif
+
+// Project includes
+#include "amgcl_raw_solver.h"
+#include "spaces/ublas_space.h"
+#include "utilities/profiler.h"
+
+// External includes
+#include "amgcl/adapter/ublas.hpp"
+#include "amgcl/adapter/zero_copy.hpp"
+#include "amgcl/backend/builtin.hpp"
+#include "amgcl/value_type/static_matrix.hpp"
+#include "amgcl/make_solver.hpp"
+#include "amgcl/make_block_solver.hpp"
+#include "amgcl/solver/runtime.hpp"
+#include "amgcl/preconditioner/runtime.hpp"
+#include "boost/property_tree/ptree.hpp"
+#include "boost/property_tree/json_parser.hpp"
+
+#ifdef AMGCL_GPGPU
+#include <amgcl/backend/vexcl.hpp>
+#include <amgcl/backend/vexcl_static_matrix.hpp>
+#endif
+
+// STL includes
+#include <sstream> // stringstream
+#include <optional> // optional
+#include <variant> // variant
+
+
+namespace Kratos {
+
+
+#ifdef AMGCL_GPGPU
+vex::Context& GetVexCLContext() {
+    static vex::Context ctx(vex::Filter::Env);
+    [[maybe_unused]] static bool run_once = [](){
+        return true;
+    }();
+    return ctx;
+}
+
+template <int TBlockSize>
+void RegisterVexCLStaticMatrixType() {
```

Comment:
Should not this be also parameterized on value type (`double`/`float`)?

Reply:
yep I missed this one, thank you! How come this didn't cause an issue at runtime?

Reply:
could it be you did not test a single-precision block solver?

Reply:
I definitely tried, but maybe there's another bug we haven't spotted yet.

Looking at the VRAM usage during solution, requesting a single precision run does result in 20-25% less memory usage, which is mostly what I'd expect.

Reply:
You could run the tests with `export VEXCL_SHOW_KERNELS=1`, then vexcl would dump the generated kernels (the ones it actually uses) to stdout.